File: lib/autoc/set.rb
# frozen_string_literal: true


require 'autoc/collection'


module AutoC


  class Set < Collection

    def orderable? = false # No idea how to compute the ordering of this container

    def initialize(*args, set_operations: true, **kws)
      super(*args, **kws)
      @set_operations = set_operations # Instruct to emit standard set operations (conjunction, disjunction etc.)
    end

  private
    
    def configure
      super
      method(:int, :put, { target: rvalue, value: element.const_rvalue }, constraint:-> { element.copyable? && element.comparable? }).configure do
        header %{
          @brief Put a value

          @param[in] target set to put into
          @param[in] value value to put
          @return non-zero value on successful put and zero value and zero value otherwise
          
          This function puts a *copy* of the specified `value` to the set only if there is no equavalent element in the set.
          The returned value indicates whether a new value in put or the set already contains an equivalent element.

          After call to this function the set will contain an element equivalent to the `value` which is either the already contained element or a copy of the specified `value`.
          
          The function requires the element's type to be both *comparable* and *copyable*.

          @since 2.0
        }
      end
      method(:int, :push, { target: rvalue,  value: element.const_rvalue }, constraint:-> { element.copyable? && element.comparable? }).configure do
        header %{
          @brief Force put a value

          @param[in] target set to put into
          @param[in] value value to put
          @return non-zero value if a value was replaced or zero value if a new value was inserted

          This function puts a *copy* of the specified `value` to the set replacing already contained element if the is one.
          The return value indicates whether this is a replacement of an already contained element or a put of a new element.

          After call to this function the set will contain an element equivalent to the `value`.

          The function requires the element's type to be both *comparable* and *copyable*.

          @since 2.0
        }
      end
      method(:int, :remove, { target: rvalue, value: element.const_rvalue }, constraint:-> { element.comparable? }).configure do
        header %{
          @brief Remove value

          @param[in] target set to process
          @param[in] value value to remove
          @return non-zero value on successful removal and zero value otherwise

          This function removes and destroys (the first occurrence of) the element considered equal to the specified value.
          The function returns zero value if `target` contains no such element.

          @since 2.0
        }
      end
      method(:int, :subset, { target: const_rvalue, other: const_rvalue }, constraint:-> { element.comparable? }).configure do
        header %{
          @brief Check if one set is a subset of another

          @param[in] target set to compare
          @param[in] other set to compare
          @return non-zero value if `target` set is a subset of `other` set and zero value otherwise

          @since 2.0
        }
      end
      equal.configure do
        code %{
          assert(left);
          assert(right);
          return #{subset.(left, right)} && #{subset.(right, left)};
        }
      end
      method(:int, :disjoint, { left: const_rvalue, right: const_rvalue }, constraint:-> { @set_operations && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(left);
          assert(right);
          for(r = #{range.new.(left)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(#{contains.(right, '*e')}) return 0;
          }
          for(r = #{range.new.(right)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(#{contains.(left, '*e')}) return 0;
          }
          return 1;
        }
        header %{
          @brief Scan two sets for common elements

          @param[in] left set to check
          @param[in] right set to check
          @return non-zero value if sets share no common elements and zero value otherwise

          This function returns non-zero value if the specified sets are disjoint that is they share no common elements and zero value otherwise.

          @since 2.0
        }
      end
      method(:void, :join, { target: rvalue, source: const_rvalue }, constraint:-> { @set_operations && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(source);
          for(r = #{range.new.(source)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            #{put.(target, '*e')};
          }
        }
        header %{
          @brief Join sets

          @param[in] target set to merge to
          @param[in] source set to get elements from

          This function copies all (new) elements from `source` set to `target` set.
          This is effectively a set join operation _target | source -> target_.

          @since 2.0
        }
      end
      method(:void, :create_join, { target: lvalue, left: const_rvalue, right: const_rvalue }, constraint:-> { @set_operations && default_constructible? && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(left);
          assert(right);
          #{default_create.(target)}; /* TODO create a set with specified capacity right away */
          for(r = #{range.new.(left)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            #{put.(target, '*e')};
          }
          for(r = #{range.new.(right)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            #{put.(target, '*e')};
          }
        }
        header %{
          @brief Join sets into new set

          @param[out] target set to merge set into
          @param[in] left set to merge
          @param[in] right set to merge

          This function creates a new `target` set which contains copies of elements from both source sets.
          This is effectively a set join operation _left | right -> target_.

          @since 2.0
        }
      end
      method(:void, :subtract, { target: rvalue, source: const_rvalue }, constraint:-> { @set_operations && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(source);
          for(r = #{range.new.(source)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            #{remove.(target, '*e')};
          }
        }
        header %{
          @brief Set substraction

          @param[in] target set to subtract from
          @param[in] source set to get elements for subtraction

          This function removes all elements from `target` set which are contained in `source` set.
          This is effectively a set subtraction operation _target \\ source -> target_.

          @since 2.0
        }
      end
      method(:void, :create_difference, { target: lvalue, left: const_rvalue, right: const_rvalue }, constraint:-> { @set_operations && default_constructible? && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(left);
          assert(right);
          #{default_create.(target)};
          for(r = #{range.new.(left)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(!#{contains.(right, '*e')}) #{put.(target, '*e')};
          }
        }
        header %{
          @brief Perform sets substraction into new set

          @param[out] target set to contain difference
          @param[in] left set to subtract from
          @param[in] right set to subtract

          This function creates new set and populates it with the copies of values from `left` set not contained in `right` set.
          This is effectively a set subtraction operation _left \\ right -> target_.

          @since 2.0
        }
      end
      method(:void, :intersect, { target: rvalue, source: const_rvalue }, constraint:-> { @set_operations && element.copyable? && element.comparable? }).configure do
        code %{
          #{type} t;
          assert(target);
          assert(source);
          #{copy.(:t, target)};
          #{subtract.(:t, source)};
          #{subtract.(target, :t)};
          #{destroy.(:t)};
        }
        header %{
          @brief Set intersection

          @param[in] target set to retain elements
          @param[in] source set to get elements for consideration

          This function retains elements in `target` set which are also contained in `source` set.
          This is effectively a set intersection operation _target & source -> target_.

          @since 2.0
        }
      end
      method(:void, :create_intersection, { target: lvalue, left: const_rvalue, right: const_rvalue }, constraint:-> { @set_operations && default_constructible? && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(left);
          assert(right);
          #{default_create.(target)};
          for(r = #{range.new.(left)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(#{contains.(right, '*e')}) #{put.(target, '*e')};
          }
          for(r = #{range.new.(right)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(#{contains.(left, '*e')}) #{put.(target, '*e')};
          }
        }
        header %{
          @brief Perform sets intersection into new set

          @param[out] target set to contain intersection
          @param[in] left set to get elements from
          @param[in] right set to get elements from

          This function creates a new set which contains copies of the elements shared by both source sets.
          This is effectively a set intersection operation left & right -> target_.

          @since 2.0
        }
      end
      method(:void, :disjoin, { target: rvalue, source: const_rvalue }, constraint:-> { @set_operations && element.copyable? && element.comparable? }).configure do
        code %{
          #{type} t;
          assert(target);
          assert(source);
          #{copy.(:t, target)};
          #{intersect.(:t, source)};
          #{join.(target, source)};
          #{subtract.(target, :t)};
          #{destroy.(:t)};
        }
        header %{
          @brief Set symmetric difference

          @param[in] target set to retain elements
          @param[in] source set to get elements for consideration

          This function collects elements in `target` which are contained in either `target` set or `source` set, but not in both.
          This is effectively a set symmetric difference operation _target ^ source -> target_.

          @since 2.0
        }
      end
      method(:void, :create_disjunction, { target: lvalue, left: const_rvalue, right: const_rvalue }, constraint:-> { @set_operations && default_constructible? && element.copyable? && element.comparable? }).configure do
        code %{
          #{range} r;
          #{element.const_lvalue} e;
          assert(target);
          assert(left);
          assert(right);
          #{default_create.(target)};
          for(r = #{range.new.(left)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(!#{contains.(right, '*e')}) #{put.(target, '*e')};
          }
          for(r = #{range.new.(right)}; !#{range.empty.(:r)}; #{range.pop_front.(:r)}) {
            e = #{range.view_front.(:r)};
            if(!#{contains.(left, '*e')}) #{put.(target, '*e')};
          }
        }
        header %{
          @brief Perform symmetric sets difference into new set

          @param[out] target set to contain difference
          @param[in] left set to get elements from
          @param[in] right set to get elements from

          This function creates a new set which contains copies of elements which are contained in either `target` set or `source` set, but not in both.
          This is effectively a set symmetric difference operation left ^ right -> target_.

          @since 2.0
        }
      end
    end

  end # Set


end

File: src/autoc/set.py
import autoc.core
import autoc.std as std
from autoc.core import inout
from autoc.module import Code
from autoc.collection import Collection


#
class Set(Collection):
  
  def __setup__(self):
    super().__setup__()
    
    self.method("int", "put", {"target": inout(self), "element": self.element}, constraint=lambda: self.element.copyable and self.element.comparable, brief="Insert element if not present",
      description="""
        @param[in,out] target the set to insert into
        @param[in] element the element to insert - ignored when the set already holds an equal element
        @return non-zero if the element was inserted and zero if an equal element was already present
      """)
    self.method("int", "remove", {"target": inout(self), "element": self.element}, constraint=lambda: self.element.comparable, brief="Remove element if present",
      description="""
        @param[in,out] target the set to remove from
        @param[in] element the element to remove
        @return non-zero if the element was removed and zero if the set held no equal element
      """)

    # The algebraic operations are composed entirely out of the protocol primitives so every
    # set implementation inherits them. They mutate the target in place and return the number
    # of the elements added, removed or changed. The bodies are built lazily because they call
    # the put/remove/contains operations which are inactive for the elements failing the traits.
    # The other set is the second operand - the self operations are shortcut
    range = self.range
    r = range.variable("r")
    temp = self.variable("temp")
    algebra_constraint = lambda: self.element.copyable and self.element.comparable

    with self.method("int", "union", {"target": inout(self), "other": self}, constraint=algebra_constraint, brief="Add all elements from other set",
      description="""
        @param[in,out] target the set to add the elements to
        @param[in] other the set whose elements are added - the target itself is allowed and keeps it unchanged
        @return the number of elements added
      """) as f:
      f.code = lambda f=f: f"""
        size_t added;
        {r.definition};
        assert(target);
        assert({f.other});
        if(target == {f.other}) return 0;
        added = 0;
        for({r} = {range.new(f.other)}; !{range.empty(r)}; {range.move_front(r)}) {{
          added += {self.put(f.target, range.front_view(r))};
        }}
        return added;
      """

    with self.method("int", "difference", {"target": inout(self), "other": self}, constraint=algebra_constraint, brief="Remove all elements found in other set",
      description="""
        @param[in,out] target the set to remove the elements from
        @param[in] other the set whose elements are removed - the target itself is allowed and empties it in place
        @return the number of elements removed
      """) as f:
      f.code = lambda f=f: f"""
        size_t removed;
        {r.definition};
        assert(target);
        assert({f.other});
        if(target == {f.other}) {{
          removed = {self.size(f.target)};
          {self.destroy(f.target)};
          {self.create(f.target)};
          return removed;
        }}
        removed = 0;
        for({r} = {range.new(f.other)}; !{range.empty(r)}; {range.move_front(r)}) {{
          removed += {self.remove(f.target, range.front_view(r))};
        }}
        return removed;
      """

    with self.method("int", "intersection", {"target": inout(self), "other": self}, constraint=algebra_constraint, brief="Keep only elements also present in other set",
      description="""
        @param[in,out] target the set to keep the elements in
        @param[in] other the set to intersect with - the target itself is allowed and keeps it unchanged
        @return the number of elements removed
      """) as f:
      f.code = lambda f=f: f"""
        size_t removed;
        {r.definition};
        {temp.definition};
        assert(target);
        assert({f.other});
        if(target == {f.other}) return 0; /* nothing is removed intersecting with self */
        {self.create(temp)};
        {self.copy(temp, f.target)};
        removed = 0;
        for({r} = {range.new(temp)}; !{range.empty(r)}; {range.move_front(r)}) {{
          if(!{self.contains(f.other, range.front_view(r))}) removed += {self.remove(f.target, range.front_view(r))};
        }}
        {self.destroy(temp)};
        return removed;
      """

    with self.method("int", ("symmetric", "difference"), {"target": inout(self), "other": self}, constraint=algebra_constraint, brief="Remove elements in both sets, add elements in only one",
      description="""
        @param[in,out] target the set to change
        @param[in] other the set to change the target by - the target itself is allowed and empties it in place
        @return the number of elements added or removed
      """) as f:
      f.code = lambda f=f: f"""
        size_t changed;
        {r.definition};
        assert(target);
        assert({f.other});
        if(target == {f.other}) {{
          changed = {self.size(f.target)};
          {self.destroy(f.target)};
          {self.create(f.target)};
          return changed;
        }}
        changed = 0;
        for({r} = {range.new(f.other)}; !{range.empty(r)}; {range.move_front(r)}) {{
          if({self.contains(f.target, range.front_view(r))}) changed += {self.remove(f.target, range.front_view(r))};
          else changed += {self.put(f.target, range.front_view(r))};
        }}
        return changed;
      """

    with self.method("int", ("is", "subset"), {"target": self, "other": self}, constraint=lambda: self.element.comparable, brief="Check if all elements are in other set",
      description="""
        @param[in] target the set to test
        @param[in] other the set to test against
        @return non-zero if every element of the target is also present in the other set
      """) as f:
      f.code = lambda f=f: f"""
        {r.definition};
        assert(target);
        assert({f.other});
        for({r} = {range.new(f.target)}; !{range.empty(r)}; {range.move_front(r)}) {{
          if(!{self.contains(f.other, range.front_view(r))}) return 0;
        }}
        return 1;
      """

    with self.method("int", ("is", "superset"), {"target": self, "other": self}, constraint=lambda: self.element.comparable, brief="Check if all elements of other are in this set",
      description="""
        @param[in] target the set to test against
        @param[in] other the set to test
        @return non-zero if every element of the other set is also present in the target
      """) as f:
      f.code = lambda f=f: f"""
        {r.definition};
        assert(target);
        assert({f.other});
        for({r} = {range.new(f.other)}; !{range.empty(r)}; {range.move_front(r)}) {{
          if(!{self.contains(f.target, range.front_view(r))}) return 0;
        }}
        return 1;
      """

    # TODO pop, ...


#
_ceil_power2 = Code(dependencies=(std.size_t, autoc.core._linkage_code), definitions="""
  AUTOC_EXTERN
  size_t _autoc_ceil_power2(size_t value);
""", implementation="""
  size_t _autoc_ceil_power2(size_t value) {
    if(value == 0) return 1;
    --value;
    value |= value >> 1;
    value |= value >> 2;
    value |= value >> 4;
    value |= value >> 8;
    value |= value >> 16;
    if(sizeof(size_t) >= 8) value |= value >> 32;
    return ++value;
  }
""")